content/blog/(best-practices)/sealos-higress-ingress-performance-optimization-minute-to-second/index.md

```diff
@@ -0,0 +1,240 @@
+
+### The Filterchain Serialization Nightmare
+
+Envoy's flame graphs revealed a shocking truth: the vast majority of processing time was consumed by protobuf serialization. To understand why, we dove deep into Envoy's change detection mechanism.
+
+![Envoy Ingress change flame graph](./images/envoy-ingress-change-flamegraph-protobuf.png)
+_Runtime flame graph showing protobuf serialization dominating CPU usage_
+
+Here's the elegant problem: How does Envoy identify which filterchains have changed among 10,000 entries when:
+
+- Each filterchain lacks a dedicated key field
+- Complete content comparison is necessary to detect modifications
+- The controller-gateway system provides no predictable ordering
+
+The original Envoy implementation serializes the entire filterchain configuration and uses that as a hashmap key for lookups. For the hash function, they chose [xxhash](https://github.com/Cyan4973/xxHash) because of its speed - but even fast hashing becomes expensive when applied to massive serialized configurations.
+
+![Envoy hashing serialized filterchain content with xxhash](./images/envoy-filterchain-xxhash-code.png)
+_Original Envoy implementation using expensive serialization for hashing_
+
+### The Hash Function Horror Show
+
+The situation was worse than we initially realized. Envoy's use of [`absl::flat_hash_map`](https://abseil.io/docs/cpp/guides/container#hash-tables) meant hash functions were called repeatedly - not just for lookups, but for assignments and updates. With 20,000+ filterchains, these seemingly innocent operations multiplied into performance armageddon.
+
+![Envoy's absl::flat_hash_map repeatedly calling hash functions](./images/envoy-absl-flat-hash-map-code.png)
+_The hash map implementation causing repeated expensive hash calculations_
+
+## Act III: The Breakthrough Solution
+
+Our solution required surgical precision across multiple system layers:
+
+### Revolutionary Recursive Hashing
+
+We abandoned serialization entirely, implementing recursive hashing where a type's hash value is calculated from its subfields' hash values. This eliminated expensive string formatting and memory allocation cycles.
+
+### Strategic Hash Caching
+
+We implemented comprehensive hash caching, even modifying [`Protobuf::Message`](https://protobuf.dev/) assignment methods to enable cache copying. This prevented redundant hash calculations across system layers.
+
+### The CachedMessageUtil Innovation
+
+Our `CachedMessageUtil` class, built upon Protobuf foundations, minimized Envoy modifications while maximizing performance gains. This approach reduced maintenance overhead for future upgrades while delivering dramatic improvements.
+
+![Envoy MessageUtil implements Protobuf hash value caching](./images/envoy-cachedmessageutil-hash-cache-1.png)
+
+![Envoy MessageUtil implements Protobuf hash value caching](./images/envoy-cachedmessageutil-hash-cache-2.png)
```

Comment:
Before:
> ![Envoy MessageUtil implements Protobuf hash value caching](./images/envoy-cachedmessageutil-hash-cache-2.png)

After:
![Our hash function v.s. the original hash function](./images/envoy-cachedmessageutil-hash-cache-2.png)